Go: Promote `go/hardcoded-key` from experimental

atorralba opened the pull request:
Promotes the query `go/hardcoded-key` from experimental, merging it with `go/hardcoded-credentials` for consistency with other languages. Where possible, plain CodeQL models have been migrated to models-as-data.

atorralba commented on go/ql/lib/semmle/go/security/Jwt.qll:
```diff
@@ -0,0 +1,44 @@
+private class SquareJoseKey extends HardcodedCredentials::Sink {
+  SquareJoseKey() {
+    exists(Field f, string pkg |
+      pkg = ["github.com/square/go-jose/v3", "gopkg.in/square/go-jose.v2"]
```
Should I use the `package` predicate here instead of hardcoding `v3`? I'm not sure it would work in the case of `go-jose.v2`, though.

owen-mc replied:
Only use `package()` if you want to cover all versions. It's fine to not use it if you are specifying one particular version of the package.

atorralba replied:
```suggestion
      pkg = [package("github.com/square/go-jose", ""), "gopkg.in/square/go-jose.v2"]
```

smowton commented on go/ql/lib/semmle/go/security/HardcodedCredentials.qll:
```diff
@@ -0,0 +1,158 @@
+
+  /** A hardcoded string literal as a source for hardcoded credentials. */
+  private class HardcodedStringSource extends Source {
+    HardcodedStringSource() { this.asExpr() instanceof StringLit }
```
I note a few other queries / languages have some exclusions for obviously non-credential strings (empty string, password123, etc)-- as a followup, worth considering adopting similar here.

atorralba replied:
Good point as well! Done in https://github.com/github/codeql/pull/15527/commits/2a3dfbe99f254d0ed42f10c6eb19f05f8c3db71a.

smowton commented on go/ql/lib/semmle/go/security/HardcodedCredentials.qll:
```diff
@@ -0,0 +1,158 @@
+
+  /**
+   * Holds if the guard `g` in its branch `branch` validates the expression `e`
+   * by comparing it to a disallowed literal.
```
```suggestion
   * by comparing it to a literal.
```

smowton commented on go/ql/lib/semmle/go/security/HardcodedCredentials.qll:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  /**
+   * A value validated by comparing it to a disallowed constant value.
```
```suggestion
   * A value validated by comparing it to a constant value.
```

smowton commented on go/ql/lib/semmle/go/security/HardcodedCredentials.qll:
```diff
@@ -0,0 +1,158 @@
+      exists(ControlFlow::ConditionGuardNode guard, SsaWithFields errorVar, ReturnStmt r |
+        guard.ensuresNeq(errorVar.getAUse(), Builtin::nil().getARead()) and
+        guard.dominates(this.getBasicBlock()) and
+        r.getExpr(1) = errorVar.getAUse().asExpr() and
```
Nitpick: `return value, err` is common but not universal -- consider generalising this and the above to "the one return value with error type vs. the other non-error-typed returns"

atorralba replied:
Good point. Turns out we have `DataFlow::returnedWithError`, so this can be greatly simplified: https://github.com/github/codeql/pull/15527/commits/d17defb00eae489ac3854f2fb9fab8a089906094.

smowton commented on go/ql/src/Security/CWE-798/HardcodedCredentials.ql:
```diff
@@ -31,22 +32,35 @@ predicate isSensitive(DataFlow::Node sink, SensitiveExpr::Classification type) {
+predicate hardcodedPrivateKey(DataFlow::Node node, SensitiveExpr::Classification type) {
+  node.getStringValue()
+      .regexpMatch("(?s)-+BEGIN\\b.*\\bPRIVATE KEY-+.+-+END\\b.*\\bPRIVATE KEY-+\n?") and
+  (node.asExpr() instanceof StringLit or node.asExpr() instanceof AddExpr) and
```
Is there anything with a `getStringValue` and that matches that regex we wouldn't want to consider?

atorralba replied:
Do you mean that `(node.asExpr() instanceof StringLit or node.asExpr() instanceof AddExpr)` is not necessary? If so, yes, you're probably right, `getStringValue` already restricts it enough.

```suggestion
```

smowton commented on go/ql/test/query-tests/Security/CWE-798/jwt.go:
```diff
@@ -139,60 +138,56 @@ func gogfjwt() interface{} {
+	safeName := []byte("key12")
 	token := iris.NewTokenWithClaims(nil, nil)
-	tokenString, _ := token.SignedString(mySecret)
+	tokenString, _ := token.SignedString(safeName) // BAD
```
safeName suggests this isn't in fact bad?

atorralba replied:
I renamed `mySecret` to `safeName` because I didn't want the other part of the query (`sensitiveAssignment`) to catch it. But independently of the name, a hardcoded string flowing to `SignedString` is a problem, and that's what I want to test.

atorralba replied:
I opted for using a more neutral name (`key`) which is ambiguous enough to not be picked up by the heuristic and less confusing at the same time: https://github.com/github/codeql/pull/15615

atorralba commented:
@smowton @owen-mc thanks for your reviews, I think I addressed all your comments. I also split things a bit better into framework libraries (instead of having it all in `Jwt.qll`, which doesn't seem to be the norm in the Go codebase and aligns better with https://github.com/github/codeql/pull/15585), and rebased to fix a conflict.

smowton reviewed (approved):
One note re: naming in the tests (followup PR?) otherwise looks good
